Add wallbox integration

hesselonline opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

I've added a component that interacts with the myWallbox portal (my dot wallbox dot com). This is a portal for an EV charging wallbox, using this component you can see the data of the wallbox. Also, it allows you to pause charging and lock/unlock the wallbox. The component uses an external library (wallbox) to interact with the API, it also uses the new config flow to be configured (so no configuration (yaml) sample below).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X ] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Wallbox component added

Commit:
Merge remote-tracking branch 'upstream/dev' into dev

Commit:
resolved mergeconflicts from upstream

Commit:
fixed an incorrect removal in CODEOWNERS file

homeassistant commented:
Hello @hesselonline,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`7177411f8f805099aa4a02a97b69878685983461`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;48082&#x2F;commits&#x2F;7177411f8f805099aa4a02a97b69878685983461) - This commit has something that looks like an email address (hessel.vanes@kpn.com). Maybe try linking that to GitHub?.
- [`d36cb3084e9587b1721f7435a3a3937185967159`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;48082&#x2F;commits&#x2F;d36cb3084e9587b1721f7435a3a3937185967159) - This commit has something that looks like an email address (hessel.vanes@kpn.com). Maybe try linking that to GitHub?.
- [`d5348263ee6f0798bad26e25713b9c0eaa9c92f0`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;48082&#x2F;commits&#x2F;d5348263ee6f0798bad26e25713b9c0eaa9c92f0) - This commit has something that looks like an email address (hessel.vanes@kpn.com). Maybe try linking that to GitHub?.
- [`dc173942e1dec2a63acc488b07abe377e66dcda8`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;48082&#x2F;commits&#x2F;dc173942e1dec2a63acc488b07abe377e66dcda8) - This commit has something that looks like an email address (hessel.vanes@kpn.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Commit:
fixes for pullrequest automatic test

homeassistant commented:
Hi @hesselonline, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#48082).

Once you do that we will be able to review and accept this pull request.

Thanks!

raz0rf0x reviewed (changes requested):
Remove linefeed from line 1 to bring docstring to compliance in homeassistant/components/wallbox/switch.py.

raz0rf0x commented on homeassistant/components/wallbox/switch.py:
```diff
@@ -0,0 +1,139 @@
+""""Home Assistant component for accessing the Wallbox Portal API, the switch component allows pausing/resuming and lock/unlock.
+    """
```
Remove

raz0rf0x commented on homeassistant/components/wallbox/switch.py:
```diff
@@ -0,0 +1,139 @@
+""""Home Assistant component for accessing the Wallbox Portal API, the switch component allows pausing/resuming and lock/unlock.
```
Add `"""` to end. Fix triple quotes at beginning.

hesselonline replied:
Thnx, done.

Commit:
clean up code after PR tests

Commit:
fixed strings.json

Commit:
fix config_flow error > wallbox

Commit:
fixed some formatting issues

Commit:
fix pylint warnings

Commit:
fixed error in number.py > set value

Commit:
pylint warnings fixed

Commit:
some more pylint fixes

Commit:
isort fixes

Commit:
fix unused_import pylint

Commit:
remove tests

Commit:
remove test requirements

Commit:
config flow test

Commit:
test errors resolved

Commit:
test file formatting

Commit:
isort on test file

Commit:
sensor test

Commit:
isort on test

Commit:
isort test const

Commit:
remove not working sensor test

Commit:
remove test const

Commit:
add switch, number and lock test

Commit:
docstrings for test classes

Commit:
sort test_number, create test_sensor

Commit:
additional tests

Commit:
fix test error

frenck reviewed (commented):
Please limit the size of this PR by adding only a single platform.
More platforms can be added once this PR has been merged.

See: <https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible>


Additionally, there is a file that doesn't belong in this PR? `.coverage.9bd5630fcc84.19872.224949`

Commit:
reduced PR to 1 component

Commit:
newline in const

Commit:
ignore test coverage -> dependency on external device (wallbox)

Commit:
do not ignore config_flow

Commit:
add test for validate_input

Commit:
remove obsolete import

Commit:
additional test config flow

Commit:
change test sensor

Commit:
docstring

Commit:
add additional test for exceptions

Commit:
fix test_config

Commit:
more tests

Commit:
fix test_config_flow

Commit:
fixed http error test

Commit:
catch connectionerror and introduce testing for this error

Commit:
remove .coveragefile

Commit:
change comment

hesselonline commented:
I removed the strange file (I think it came from trying to determine test coverage). Also, for this PR I limited it to a sensor component only. Will extend to other platforms in future PR's

zbuh commented:
is this ready for merge?
any blocking points to be fixed?

hesselonline commented:
I think it is ready for merging

janiversen commented:
Before it can be merged, you need to solve the merge conflict. Apart from that there is a request for review outstanding.

hesselonline commented:
yeah sure, but the merge conflict wasn't there when the PR was created. Its just that it has been sitting here for weeks... So, I can resolve a merge conflict but I'm not sure if that will create any momentum on this PR.

janiversen commented:
We all have the same issue with merge conflicts, especially in integrations with many changes. I was just giving a polite advice.

hesselonline commented:
I understand that merge conflicts come with the territory, and I don't mean to be impolite (sorry if I came across that way). Its just that I'm seriously wondering if its worth the effort for me to fix this merge conflict.

janiversen commented:
Let me first say, some PR wait months, for various reasons but one of the important one are the amount of PR which have to be reviewed.

I sure understand your argument and yes there are no guarantees in this world. I looked at your PR this morning, started a review and left a comment, the rest is your decision.

hesselonline committed:
Merge branch 'dev' into dev

hesselonline commented:
I think this fixes the merge conflict

janiversen commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        "station": str,
```
Please make the keys constants and refer to e.g. STATION and not "station".

hesselonline replied:
OK, changed this to use constants from .const and homeassistant.const

janiversen replied:
I cannot see it.

hesselonline replied:
no, did this locally, will make a big commit when I have processed more changes

janiversen replied:
ok, and seen from that POW it is totally correct not to resolve the conversation

janiversen commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+    def authenticate(self) -> bool:
+        """Authenticate using Wallbox API."""
+        try:
+            wallbox = Wallbox(self._username, self._password)
```
Seems like a duplicate of "async_setup_entry" ?

hesselonline replied:
not completely, but to some extend I guess. This is mainly based on the scaffolding component for the config flow. Here I try to validate input in the config flow -> prevent that the user enters incorrect data and ends up with a non-working integration. To my understanding the config flow part is separate from the init file, so there is some duplicate work here.

janiversen replied:
Yes your understanding is correct..the init file is used when your integration start up, and config_flow when a user want to configure. I guess I did not write clear enough, duplicate code is more likely to grow different over time, so it is better to have a common function.

janiversen commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+    def get_data(self) -> bool:
+        """Get new sensor data for Wallbox component."""
+
+        try:
```
I suggest to make a function with the block and a function object as parameter to call e.g. getChargerStatus

You might also be able to centralise the add_executor_job, and thus be free to make many of the methods async.

hesselonline replied:
That's a good suggestion, that would allow me to remove some duplicate code and make it more clean. I'm going to write that down as a suggestion for a next release if that's ok with you.

janiversen replied:
We can park "make a function" to a later PR, that is no problem. But your distributed add_executor_job is a concern. I know you need to use them (I do the same in modbus/modbus.py), but please try to keep as close to the 3rd party library as possible and preference is to isolate them in WallboxHub. Doing makes the bigger part of your code clean and nice async.

janiversen commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+)
+
+
+class PlaceholderHub:
```
```suggestion
class WallboxHub:
```

hesselonline replied:
Done

janiversen replied:
Thanks, normally when you consider a request done, hit "Resolve conversations" as well. Of course if you answer a request with a comment, then please do not resolve the conversation 😄

hesselonline replied:
yeah, but I did it locally, so marking it as resolved would not be fully accurate as of now ;-)

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,99 @@
+    async def async_update_data():
+
+        try:
+            return await hass.async_add_executor_job(wallbox_updater, wallbox, station)
```
It would be nice it wallbox_updater was async, so that you had the async_add_executor_job centralised in one py file, which makes it easier to update in the future.

hesselonline replied:
yes, I can make that change here.

janiversen replied:
👍

janiversen commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -0,0 +1,157 @@
+    )
+
+    with patch(
+        "homeassistant.components.wallbox.config_flow.PlaceholderHub.authenticate",
```
I suggest to mock the library function so PlaceHolderHub.authenticate() gets tested.

janiversen commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,48 @@
+
+    with patch("wallbox.Wallbox", return_value=True):
+        assert wallbox.async_setup_entry(hass, MOCK_CONFIG)
+
```
Please include the negative test as well "return_value=False"

The idea being that this invokes another part of the code.

janiversen reviewed (changes requested):
A first review.

In general it looks good.

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,63 @@
+    )
+
+    try:
+        await hass.async_add_executor_job(wallbox.authenticate)
```
No, it just uses a class method to authenticate the instance of the wallbox class with the api. It returns none.

janiversen replied:
Looking at the method it also raises InvalidAuth, which you do not catch. Please look at all places where you call wallbox.authenticate().

hesselonline committed:
Update homeassistant/components/wallbox/__init__.py

review suggestion by janiversen

Co-authored-by: jan iversen <jancasacondor@gmail.com>

hesselonline committed:
Update homeassistant/components/wallbox/__init__.py

review suggestion by janiversen (format only)

Co-authored-by: jan iversen <jancasacondor@gmail.com>

janiversen reviewed (changes requested):
Looking forward to see the rest of your changes. Please add a comment mentioning my user, when you want a new review.

Commit:
Processed review comments, include more testing for sensor component

Commit:
Merge branch 'dev' of github.com:hesselonline/core into dev

janiversen reviewed (changes requested):
Looks better, only a few requests to solve.

Commit:
Isolated the async_add_executor_job to make the solution more async

hesselonline commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,92 @@
+        name="wallbox",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=30),
```
done

hesselonline commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,92 @@
+    filtered_data = {k: data[k] for k in SENSOR_TYPES if k in data}
+
+    for key, value in filtered_data.items():
+        sensor_round = SENSOR_TYPES[key]["ATTR_ROUND"]
```
I used another prefix for these items

hesselonline commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,92 @@
+
+async def wallbox_updater(wallbox, hass):
+    """Get new sensor data for Wallbox component."""
+    data = await hass.async_add_executor_job(wallbox.get_data)
```
Done

Commit:
add a config flow test

janiversen commented:
Please click "Resolve conversation" on the items you see as closed, so I know what you still see as open.

hesselonline commented:
@janiversen I resolved the issues. In short, I moved the WallboxHub part to __init__ and all the async stuff is now only there. Had to fix the testing because of that, and also included more tests / tests for the previously excluded files.

janiversen commented:
You have 1.321 files changes ????? could be a merge gone wrong.

I normally do:
```
git checkout dev
git pull
git checkout <branch>
git rebase dev
(solve any merge issues, however you should not have any).
git push -f
```

Commit:
Revert "add a config flow test"

This reverts commit 9c1af82fffeb0b46f11ada1000e19b66fd5fd0f1.

Commit:
Revert "Isolated the async_add_executor_job to make the solution more async"

This reverts commit 0bf034c3318f27e649389830d4ad7a7e10eb2d6f.

Commit:
Make component more async and add config flow tests

hesselonline commented:
@janiversen sorry, my bad. The pre-commit hooks don't work outside my DevContainer setup. So I tried a git add / commit in the container, apparently this is the result. But, of course, I should have checked the affected files of the add function...
I reverted and re-committed the change.

janiversen commented:
Making errors is something very normal, not learning from them is bad.

janiversen commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,112 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Wallbox from a config entry."""
+    wallbox = WallboxHub(
```
```suggestion
    self.wallbox = WallboxHub(
```
why not reuse ?

hesselonline replied:
that would not work there, its not inside the class. Do you mean something different?

janiversen replied:
ups my bad

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,93 @@
+    filtered_data = {k: data[k] for k in SENSOR_TYPES if k in data}
+
+    for key, value in filtered_data.items():
+        sensor_round = SENSOR_TYPES[key]["ST_ROUND"]
```
Please use a const.

janiversen commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,93 @@
+        """Initialize a Wallbox sensor."""
+        super().__init__(coordinator)
+        self._properties = SENSOR_TYPES[ent]
+        self._name = f"{config.title} {self._properties['ST_LABEL']}"
```
Plase use consts.

janiversen commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -0,0 +1,128 @@
+async def test_validate_input(hass):
+    """Test we can validate input."""
+    data = {
+        "station": "12345",
```
station/username/password is used many time please change to const (homeassistant 
.const probably have sone of them)

janiversen reviewed (changes requested):
Small issues, you have changed the bigger issues in a nice way.

bdraco commented on homeassistant/components/wallbox/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+    """Handle a config flow for Wallbox."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
```suggestion
```

This has moved to `manifest.json`

janiversen replied:
What happened to this suggestion ?

bdraco commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,93 @@
+        update_interval=timedelta(seconds=UPDATE_INTERVAL),
+    )
+
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

bdraco replied:
But move this into `__init__.py` and store the coordinator in `hass.data`

hesselonline replied:
I had to google the 'first refresh' function, but that seems like a good addition. What would be the benefit of moving this to init? And where would you store the coordinator? Currently I store the wallbox object there, which has the benefit that I can include additional methods for future functionality

bdraco replied:
Make hass.data[DOMAIN][entry.entry_id] as dict

hesselonline replied:
Thnx, I looked at the Roku implementation. I think that is quite a nice way to do it, but this would also mean a substantial rewrite (because I would also need a custom coordinator class if I want to use additional methods for the api). So I'm going to write this one down on my wishlist for a next big change if that's ok with you.

bdraco replied:
Thats fine to not use the Roku method.  It still needs to be moved into `__init__` or Home Assistant won't be able to retry the setup since it needs to happen before the platforms are loaded

hesselonline replied:
Okay, I moved it around a bit. The coordinator is now being created in __init__ and becomes part of the WallboxHub class (which in turn is being stored in hass.data[DOMAIN][entry.entry_id]).

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    async def async_authenticate(self, hass) -> bool:
+        """Authenticate using Wallbox API."""
+        try:
+            wallbox = Wallbox(self._username, self._password)
```
OK, what do you mean exactly, because this already is a method in WallboxHub? Or do you mean to create a 'private' method which is not async but does the try catch part and then a public method to call that? Or do you mean something entirely different?

janiversen replied:
I mean keep async_authenticate in WallboxHub as the public interface. The try/except should not contain async_add_executor_job

new method 
def _authenticate():
       try:
            wallbox = WallBox …. the library
            wallbox.authenticate()
       except

and in async_authenticate
         await hass.async_add_executor_job(self._authenticate)

that try/except and library call run in the same thread.

hesselonline replied:
yes, I understand now. Done

hesselonline commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    async def async_get_data(self, hass) -> bool:
+        """Get new sensor data for Wallbox component."""
+        try:
+            wallbox = Wallbox(self._username, self._password)
```
Again, I do not fully understand. But let me explain my logic here. I want people to be able to create multiple instances of this platform/integration, because they might have multiple stations which might use multiple logins (a few people on the forum already mention using different stations). So, for me a wallbox is the combination of Station/User/Password. In setup_entry I instantiate an instance of WallboxHub and pass it in hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id].

janiversen replied:
If they create multiple instances of your platform async_setup_entry will be called multiple tines, meaning you have 1 instance of WallBoxHub for each username/password. So it would be natural to store the library object WallBox in WallBoxHub.

hesselonline replied:
When working on your previous comment I actually got to the same conclusion (ended up storing wallbox in a hub variable) and I think that is exactly what you mean here. So, good suggestion and done.

hesselonline committed:
Changes based on review comments

hesselonline committed:
made _ methods in WallboxHub for the 'non-async' call to the API and try-catch. Stored the wallbox in the class.

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        data = await hass.async_add_executor_job(self._get_data)
+        return data
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Wallbox component."""
+    return True
+
+
```
```suggestion
```

Not needed

bdraco commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,101 @@
+        update_interval=timedelta(seconds=UPDATE_INTERVAL),
+    )
+
+    await coordinator.async_config_entry_first_refresh()
```
The first update needs to happen in `__init__`'s `async_setup_entry` before the platform is loaded

hesselonline committed:
moved the coordinator to __init__ and pass it as part of the WallboxHub class

hesselonline committed:
removed obsolete function in __init__

janiversen commented:
This PR is ready to be approved seen from my side, but I think there is one request from @bdraco that have not been solved (remove "CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL" since it is in manifest.json)

@bdraco if your requests are resolved, I will approve.

hesselonline committed:
removed CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL

hesselonline commented:
@janiversen and @bdraco, it seems I forgot to delete those rows in config_flow.py (although I did add them to the manifest.json). I have changed this now.

janiversen reviewed (approved):
Looks good.

I will let it hand in approved state a while in case @bdraco have some objections.

janiversen commented:
Seem you have a little misspelling, please fix that.

hesselonline committed:
fixed spelling and imports on test files

hesselonline commented:
OK, I think I have all the inconsistencies fixed in the test files now. One thing that confuses me, I thought we needed a version in the manifest (read it somewhere), so I added it before. But now that gives me an error in the hassfest check (so, now I removed it again).

janiversen commented:
The version in manifest.json is for custom components, not fir core components. A bit confusing.

janiversen commented:
an iSort error, that is unusual, looks like you do not use our standard githooks (gets installed with the development environment and are highly recommended). You need to sort the imports.

hesselonline committed:
did isort on component files

hesselonline commented:
Did an additional isort on the component files. To answer your question, I'm running a windows environment with the VSCode DevContainer. The pre-commit hooks don't work for me, so I'm forced to skip them (maybe i can get them working in the WSL, will try later)

hesselonline committed:
Merge branch 'dev' of https://github.com/home-assistant/core into dev

hesselonline commented:
BTW got it working now in WSL2, had to do somethings different compared to the guide (https://developers.home-assistant.io/docs/development_environment/), might be good to include that there (I can make a proposal if necessary).

janiversen commented:
Proposals are very welcome. Restarted CI, lets hope for green light.

hesselonline commented:
it seems to fail in a different component:
Warning: homeassistant/components/sia/alarm_control_panel.py:99:4: W0006: __init__ should have explicit return type "None" (hass-constructor-return)
************* Module homeassistant.components.sia.hub
Warning: homeassistant/components/sia/hub.py:34:4: W0006: __init__ should have explicit return type "None" (hass-constructor-return)

janiversen commented:
Should be solved.

janiversen commented:
Looking forward to see more PRs from you.

hesselonline commented:
Thnx, and thnx for the assist -> learned a lot in the past few days! I will expand this integration with a few additional platforms (lock, number and switch) anything I should be aware of?

janiversen commented:
In your next PR we will focus more on the test part especially if you add files to .coveragerc.

You might receive some comments on this PR even though it is merged, because all merge commits are controlled an extra time (more eyes see more).

If you have questions, you can always submit a half-baked PR, mark it as a draft, and ask for advice (feel free to mention me).

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,147 @@
+    def __init__(self, msg=""):
+        """Create a log record."""
+        super().__init__()
+        _LOGGER.error("Cannot connect to Wallbox API. %s", msg)
```
Please move side effects out of init.  We don't want exceptions to trigger logger messages.

bdraco commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,147 @@
+    def __init__(self, msg=""):
+        """Create a log record."""
+        super().__init__()
+        _LOGGER.error("Cannot authenticate with Wallbox API. %s", msg)
```
Please move side effects out of init.  We don't want exceptions to trigger logger messages.

bdraco commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -0,0 +1,128 @@
+        side_effect=alternate_get_charger_status_method,
+    ):
+
+        result = await config_flow.validate_input(hass, data)
```
Please don't call the integration code directly, instead create a new flow with the input you want to validate similar to `test_form_cannot_connect`

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -0,0 +1,128 @@
+
+async def test_configflow_class():
+    """Test configFlow class."""
+    configflow = config_flow.ConfigFlow()
```
Please don't call the integration code directly, instead create a new flow with the input you want to validate similar to `test_form_cannot_connect`

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_config_flow.py:
```diff
@@ -0,0 +1,128 @@
+    assert configflow
+
+    with patch(
+        "homeassistant.components.wallbox.config_flow.validate_input",
```
Patch the `pypi` code, not the integration when writing tests

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+            text='{"Temperature": 100, "Location": "Toronto", "Datetime": "2020-07-23", "Units": "Celsius"}',
+            status_code=200,
+        )
+        assert await wallbox.async_setup_entry(hass, entry)
```
Set up the integration with the core interface either async_setup_component or hass.config_entries.async_setup if the integration supports config entries.

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+            text='{"jwt":"fakekeyhere","user_id":12345,"ttl":145656758,"error":false,"status":404}',
+            status_code=404,
+        )
+        assert await wallbox.async_setup_entry(hass, entry) is False
```
Set up the integration with the core interface either async_setup_component or hass.config_entries.async_setup if the integration supports config entries.

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+    """Test Wallbox Unload."""
+    hass.data[DOMAIN] = {"connections": {entry.entry_id: entry}}
+
+    assert await wallbox.async_unload_entry(hass, entry)
```
Use `assert await hass.config_entries.async_unload(config_entry.entry_id)`

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+    hass.data[DOMAIN] = {"fail_entry": entry}
+
+    with pytest.raises(KeyError):
+        await wallbox.async_unload_entry(hass, entry)
```
Use `assert await hass.config_entries.async_unload(config_entry.entry_id)`

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+            json=test_response,
+            status_code=200,
+        )
+        assert await hub.async_get_data()
```
Patch the `pypi` code and setup the integration using the standard interfaces instead of calling it directly

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+            json=test_response_rounding_error,
+            status_code=200,
+        )
+        assert await hub.async_get_data()
```
Patch the `pypi` code and setup the integration using the standard interfaces instead of calling it directly

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_sensor.py:
```diff
@@ -0,0 +1,81 @@
+    assert wallboxSensor.state
+
+
+# async def test_wallbox_updater(hass: HomeAssistantType):
+#     """Test wallbox updater."""
+#     with requests_mock.Mocker() as m:
+#         m.get(
+#             "https://api.wall-box.com/auth/token/user",
+#             text='{"jwt":"fakekeyhere","user_id":12345,"ttl":145656758,"error":false,"status":200}',
+#             status_code=200,
+#         )
+#         m.get(
+#             "https://api.wall-box.com/chargers/status/('12345',)",
+#             json=test_response,
+#             status_code=200,
+#         )
+#         await sensor.wallbox_updater(wallbox, hass)
+
+
+# async def test_wallbox_updater_rounding_error(hass: HomeAssistantType):
+#     """Test wallbox updater rounding error."""
+#     with requests_mock.Mocker() as m:
+#         m.get(
+#             "https://api.wall-box.com/auth/token/user",
+#             text='{"jwt":"fakekeyhere","user_id":12345,"ttl":145656758,"error":false,"status":200}',
+#             status_code=200,
+#         )
+#         m.get(
+#             "https://api.wall-box.com/chargers/status/('12345',)",
+#             json=test_response_rounding_error,
+#             status_code=200,
+#         )
+#         await sensor.wallbox_updater(wallbox, hass)
```
Please remove commented code

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+    with requests_mock.Mocker() as m, raises(wallbox.InvalidAuth):
+        m.get("https://api.wall-box.com/auth/token/user", text="data", status_code=403)
+
+        assert await hub.async_authenticate()
```
Patch the `pypi` code and setup the integration using the standard interfaces instead of calling it directly

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_init.py:
```diff
@@ -0,0 +1,165 @@
+    username = ("test-username",)
+    password = "test-password"
+
+    hub = wallbox.WallboxHub(station, username, password, hass)
```
Patch the `pypi` code and setup the integration using the standard interfaces instead of calling it directly

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/wallbox/test_sensor.py:
```diff
@@ -0,0 +1,81 @@
+    idx = 1
+    ent = "charging_power"
+
+    wallboxSensor = sensor.WallboxSensor(coordinator, idx, ent, entry)
```
Patch the `pypi` code and setup the integration using the standard interfaces instead of calling it directly

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco reviewed (commented):
Please address comments in as new PR.  Thanks 👍

janiversen commented:
@hesselonline  do you want help solving the new review comments ?

I did part of the old review, but did not catch the new comments.

hesselonline commented:
Thanks for the offer, I'll look at them and see if I get stuck anywhere. Quick question at this time, one of the remarks of @bdraco was about the custom exceptions in __init__ . Is there a certain agreed upon way to do exceptions/logging in HA, should I use build-in classes?

janiversen commented:
The agreed upon way is to avoid exceptions in init, typically by moving that code into async_setup()

MartinHjelmare commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,147 @@
+UPDATE_INTERVAL = 30
+
+
+class WallboxHub:
```
This can inherit `DataUpdateCoordinator`. Then we don't need to create new methods that mimic the existing coordinator methods.

MartinHjelmare commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,147 @@
+        """Authenticate using Wallbox API."""
+        return await self._hass.async_add_executor_job(self._authenticate)
+
+    async def async_get_data(self) -> bool:
```
Implement `_async_update_data` instead.

https://github.com/home-assistant/core/blob/19c9675d4a0b69799f0b606e6782fde65cfaaf88/homeassistant/helpers/update_coordinator.py#L139

MartinHjelmare replied:
The return value is not a boolean. Looks like it's a dict.

MartinHjelmare commented on homeassistant/components/wallbox/__init__.py:
```diff
@@ -0,0 +1,147 @@
+
+    def __init__(self, msg=""):
+        """Create a log record."""
+        super().__init__()
```
Normally the custom exception message is passed as an argument to super here. Remove the logging per below.

MartinHjelmare commented on homeassistant/components/wallbox/const.py:
```diff
@@ -0,0 +1,99 @@
+
+CONF_SENSOR_TYPES = {
+    "charging_power": {
+        CONF_ICON: "mdi:ev-station",
```
Please set device class for as many sensor types as possible. When setting device class for an entity we normally don't set a custom icon.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

MartinHjelmare commented on homeassistant/components/wallbox/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/wallbox",
+  "requirements": ["wallbox==0.4.4"],
+  "ssdp": [],
```
Please remove empty items.

MartinHjelmare commented on homeassistant/components/wallbox/const.py:
```diff
@@ -0,0 +1,99 @@
+        CONF_NAME: "Charging Power",
+        CONF_ROUND: 2,
+        CONF_UNIT_OF_MEASUREMENT: POWER_KILO_WATT,
+        STATE_UNAVAILABLE: False,
```
Where do we use `STATE_UNAVAILABLE`?

MartinHjelmare commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,61 @@
+    )
+
+
+class WallboxSensor(CoordinatorEntity, Entity):
```
Please inherit from `SensorEntity` besides `CoordinatorEntity`.

MartinHjelmare commented on homeassistant/components/wallbox/sensor.py:
```diff
@@ -0,0 +1,61 @@
+    def __init__(self, coordinator, idx, ent, config):
+        """Initialize a Wallbox sensor."""
+        super().__init__(coordinator)
+        self._properties = CONF_SENSOR_TYPES[ent]
```
`_properties` isn't used outside the init method so doesn't need to be an instance attribute. A local variable is enough.

MartinHjelmare commented on homeassistant/components/wallbox/strings.json:
```diff
@@ -0,0 +1,22 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
```
Unknown error string is never used.

MartinHjelmare commented on homeassistant/components/wallbox/strings.json:
```diff
@@ -0,0 +1,22 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
Already configured abort reason is never used.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
